pkg/agent/discoverycache.go

```diff
@@ -24,264 +25,428 @@ const (
-	crd := mv1.NewCredentialRequestDefinition("", agent.cfg.GetEnvironmentName()).GetKindLink()
-	url := fmt.Sprintf("%s/apis%s", agent.cfg.GetURL(), crd)
+	dc.logger.WithField("kind", "CredentialRequestDefinition").Trace("fetching CredentialRequestDefinitions and updating cache")
+
+	crd := mv1.NewCredentialRequestDefinition("", dc.envName).GetKindLink()
+	url := dc.formatResourceURL(crd)
 
-	// Update cache with published resources
-	existingCRDs := make(map[string]bool)
 	query := map[string]string{
 		apic.FieldsKey: apiServerFields,
 	}
 
-	crds, _ := GetCentralClient().GetAPIV1ResourceInstancesWithPageSize(query, url, apiServerPageSize)
+	crds, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(query, url, apiServerPageSize)
+	if err != nil {
+		return err
+	}
 
 	for _, crd := range crds {
-		agent.cacheManager.AddCredentialRequestDefinition(crd)
-		existingCRDs[crd.Metadata.ID] = true
+		if err := dc.handleResource(crd); err != nil {
+			return err
+		}
 	}
+
+	return nil
 }
 
-func (j *discoveryCache) updateAccessRequestCache() {
-	if agent.agentFeaturesCfg == nil || !agent.agentFeaturesCfg.MarketplaceProvisioningEnabled() {
-		return
+func (dc *discoveryCache) handleAccessRequest() error {
+	if !dc.isMpEnabled {
+		return nil
 	}
-	j.logger.Trace("updating access request cache")
 
-	// Update cache with published resources
-	existingAccessRequests := make(map[string]bool)
+	dc.logger.WithField("kind", "AccessRequest").Trace("fetching AccessRequests and updating cache")
+
 	query := map[string]string{
 		apic.FieldsKey: apiServerFields + "," + defs.Spec + "," + defs.ReferencesSubResource,
 	}
 
-	accessRequests, _ := GetCentralClient().GetAPIV1ResourceInstancesWithPageSize(
-		query, agent.cfg.GetEnvironmentURL()+"/accessrequests", apiServerPageSize,
+	arLink := mv1.NewAccessRequest("", dc.envName).GetKindLink()
+	url := dc.formatResourceURL(arLink)
+	accessRequests, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(
+		query, url, apiServerPageSize,
 	)
 
-	for _, accessRequest := range accessRequests {
-		ar := &mv1.AccessRequest{}
-		ar.FromInstance(accessRequest)
-		agent.cacheManager.AddAccessRequest(ar)
-		existingAccessRequests[accessRequest.Metadata.ID] = true
-		j.addSubscription(ar)
+	if err != nil {
+		return err
+	}
+
+	for _, req := range accessRequests {
+		if err := dc.handleResource(req); err != nil {
+			return err
+		}
 	}
+
+	return nil
 }
 
-func (j *discoveryCache) addSubscription(ar *mv1.AccessRequest) {
-	subscriptionName := defs.GetSubscriptionNameFromAccessRequest(ar)
-	if subscriptionName == "" {
-		return
+func (dc *discoveryCache) handleCredential() error {
+	if !dc.isMpEnabled {
+		return nil
+	}
+
+	dc.logger.WithField("kind", "Credential").Trace("fetching Credentials and updating cache")
+
+	credLink := mv1.NewCredential("", dc.envName).GetKindLink()
+	url := dc.formatResourceURL(credLink)
+
+	credentials, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(
+		nil, url, apiServerPageSize,
```

Comment:
@jcollins-axway Are there certain fields to query for on the credential like we do for other resources?

Reply:
I do not think we should limit the fields we get.